system/Validation/Validation.php

```diff
@@ -353,23 +355,33 @@ public function withRequest(RequestInterface $request): ValidationInterface
      *        'rule' => 'message'
      *    ]
      *
-     * @param string      $field
-     * @param string|null $label
-     * @param string      $rules
-     * @param array       $errors
+     * @param string       $field
+     * @param string|null  $label
+     * @param string|array $rules
+     * @param array        $errors
      *
      * @return $this
+     *
+     * @throws TypeError
      */
-    public function setRule(string $field, string $label = null, string $rules, array $errors = [])
+    public function setRule(string $field, string $label = null, $rules, array $errors = [])
     {
-        $this->rules[$field] = [
-            'label' => $label,
-            'rules' => $rules,
+        if (! is_array($rules) && ! is_string($rules)) {
+            throw new TypeError('$rules must be of type string|array');
+        }
+
+        $ruleSet = [
+            $field => [
+                'label' => $label,
+                'rules' => $rules,
+            ],
         ];
 
-        $this->customErrors = array_merge($this->customErrors, [
-            $field => $errors,
-        ]);
+        if ($errors) {
+            $ruleSet[ $field ]['errors'] = $errors;
+        }
+
+        $this->setRules($ruleSet);
```

Comment:
Requires adding old rules back here, since `setRules` overwrites all rules.
Also add a test for that.